client-spark/common/src/main/java/org/apache/spark/shuffle/celeborn/SortBasedPusher.java

```diff
@@ -339,24 +339,50 @@ private void growPointerArrayIfNecessary() throws IOException {
       }
 
       long used = inMemSorter.getMemoryUsage();
+      long requestedBytes = used / 8 * 2;
+      int allocateMemoryRetryCount = 0;
+      int maxMemoryAllocationRetry = 3;
       LongArray array = null;
-      try {
-        // could trigger spilling
-        array = allocateArray(used / 8 * 2);
-      } catch (TooLargePageException e) {
-        // The pointer array is too big to fix in a single page, spill.
-        logger.info(
-            "Pushdata in growPointerArrayIfNecessary, memory used {}",
-            Utils.bytesToString(getUsed()));
-        pushData(true);
-      } catch (SparkOutOfMemoryError rethrow) {
-        // should have trigger spilling
-        if (inMemSorter.numRecords() > 0) {
-          logger.error("OOM, unable to grow the pointer array");
-          throw rethrow;
+      boolean cont = true;
+      while (allocateMemoryRetryCount < maxMemoryAllocationRetry && cont) {
+        try {
+          // could trigger spilling
+          logger.info("asking for " + requestedBytes + " more bytes to accommodate more records");
+          array = allocateArray(used / 8 * 2);
```

Comment:
IIUC the intention is to use `requestedBytes` instead of `used / 8 * 2` here?

Reply:
oops....and checked in our internal version it is requestedBytes...just some typo when upstreaming